dll/win32/shell32/shlexec.cpp

```diff
@@ -2650,3 +2650,197 @@ HRESULT WINAPI ShellExecCmdLine(
 
     return HRESULT_FROM_WIN32(dwError);
 }
+
+/*************************************************************************
+ *                RealShellExecuteExA (SHELL32.266)
+ */
+EXTERN_C
+HINSTANCE WINAPI
+RealShellExecuteExA(
+    _In_opt_ HWND hwnd,
+    _In_opt_ LPCSTR lpOperation,
+    _In_opt_ LPCSTR lpFile,
+    _In_opt_ LPCSTR lpParameters,
+    _In_opt_ LPCSTR lpDirectory,
+    _In_opt_ LPSTR lpReturn,
+    _In_opt_ LPCSTR lpTitle,
+    _In_opt_ LPVOID lpReserved,
+    _In_ INT nCmdShow,
+    _Out_opt_ HANDLE *lphProcess,
+    _In_ DWORD dwFlags)
+{
+    SHELLEXECUTEINFOA ExecInfo;
+
+    TRACE("(%p, %s, %s, %s, %s, %p, %s, %p, %u, %p, %lu)\n",
+          hwnd, debugstr_a(lpOperation), debugstr_a(lpFile), debugstr_a(lpParameters),
+          debugstr_a(lpDirectory), lpReserved, debugstr_a(lpTitle),
+          lpReserved, nCmdShow, lphProcess, dwFlags);
+
+    ZeroMemory(&ExecInfo, sizeof(ExecInfo));
+    ExecInfo.cbSize = sizeof(ExecInfo);
+    ExecInfo.fMask = SEE_MASK_FLAG_NO_UI | SEE_MASK_UNKNOWN_0x1000;
+    ExecInfo.hwnd = hwnd;
+    ExecInfo.lpVerb = lpOperation;
+    ExecInfo.lpFile = lpFile;
+    ExecInfo.lpParameters = lpParameters;
+    ExecInfo.lpDirectory = lpDirectory;
+    ExecInfo.nShow = (WORD)nCmdShow;
+
+    if (lpReserved)
+    {
+        ExecInfo.fMask |= SEE_MASK_FLAG_SEPVDM;
+        ExecInfo.hInstApp = (HINSTANCE)lpReserved;
```

Comment:
`ExecInfo.fMask |= ...;` This value is surely `0x40000`.
And function parameter `lpReserved` should be renamed as `hInstApp`. I can rename `SEE_MASK_USE_RESERVED` as `SEE_MASK_USE_HINSTAPP`.

Reply:
iirc. I wrote either here or in mattermost some more detailed analysis and a suggestion of the renaming of flags.

Reply:
Whatever the name (yes it does have a member called "hInstApp": https://learn.microsoft.com/en-us/windows/win32/api/shellapi/ns-shellapi-shellexecuteinfoa#members ), this member is used differently depending on the flags passed to it (see the previous link):

> hInstApp (Type: HINSTANCE)
> [out] If SEE_MASK_NOCLOSEPROCESS is set and the [ShellExecuteEx](https://learn.microsoft.com/en-us/windows/desktop/api/shellapi/nf-shellapi-shellexecuteexa) call succeeds, it sets this member to a value greater than 32. If the function fails, it is set to an SE_ERR_XXX error value that indicates the cause of the failure. Although hInstApp is declared as an HINSTANCE for compatibility with 16-bit Windows applications, it is not a true HINSTANCE. It can be cast only to an int and compared to either 32 or the following SE_ERR_XXX error codes.

--> In this case it is used in this case as an _**output**_ parameter;

> `SEE_MASK_FLAG_HINST_IS_SITE` (0x08000000) 
> The hInstApp member is used to specify the IUnknown of an object that implements IServiceProvider. This object will be used as a site pointer. The site pointer is used to provide services to the ShellExecute function, the handler binding process, and invoked verb handlers.

--> In this case it is used as an _**input**_ parameter.

AND, in case the "use-reserved" undocumented flag is used (your case of interest), I showed you how it is used: as an input parameter that points to a specific control string that is parsed by kernel32, and this is used by progman.exe